sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -18,91 +18,81 @@
     this.kinesis = checkNotNull(simplifiedKinesisClient, "simplifiedKinesisClient");
-    shardIterator = checkpoint.getShardIterator(kinesis);
+    this.streamName = initialCheckpoint.getStreamName();
+    this.shardId = initialCheckpoint.getShardId();
+    this.shardIterator = initialCheckpoint.getShardIterator(kinesis);
   }
 
-  /**
-   * Returns record if there's any present.
-   * Returns absent() if there are no new records at this time in the shard.
-   */
-  public CustomOptional<KinesisRecord> next() throws TransientKinesisException {
-    readMoreIfNecessary();
-
-    if (data.isEmpty()) {
-      return CustomOptional.absent();
-    } else {
-      KinesisRecord record = data.removeFirst();
-      checkpoint = checkpoint.moveAfter(record);
-      return CustomOptional.of(record);
-    }
+  List<KinesisRecord> readNextBatch() throws TransientKinesisException {
+    GetKinesisRecordsResult response = fetchRecords();
+    LOG.debug("Fetched {} new records", response.getRecords().size());
+
+    List<KinesisRecord> filteredRecords = filter.apply(response.getRecords(), checkpoint.get());
+    millisBehindLatest.set(response.getMillisBehindLatest());
+    return filteredRecords;
   }
 
-  private void readMoreIfNecessary() throws TransientKinesisException {
-    if (data.isEmpty()) {
-      GetKinesisRecordsResult response = fetchRecords();
-      data.addAll(filter.apply(response.getRecords(), checkpoint));
+  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
+    try {
+      GetKinesisRecordsResult response = kinesis.getRecords(shardIterator, streamName, shardId);
+      shardIterator = response.getNextShardIterator();
+      return response;
+    } catch (ExpiredIteratorException e) {
+      LOG.info("Refreshing expired iterator", e);
+      shardIterator = checkpoint.get().getShardIterator(kinesis);
+      return fetchRecords();
     }
   }
 
-  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
-    GetKinesisRecordsResult response = null;
-    do {
-      try {
-        response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
-            checkpoint.getShardId());
-        shardIterator = response.getNextShardIterator();
-      } catch (ExpiredIteratorException e) {
-        LOG.info("Refreshing expired iterator", e);
-        shardIterator = checkpoint.getShardIterator(kinesis);
-      }
-    } while (response == null || gotEmptyResponseButIsBeforeEndOfTheStream(response));
-    return response;
+  ShardCheckpoint getCheckpoint() {
+    return checkpoint.get();
   }
 
-  private boolean gotEmptyResponseButIsBeforeEndOfTheStream(GetKinesisRecordsResult response) {
-    return response.getRecords().isEmpty() && response.getMillisBehindLatest() > 0;
+  boolean isUpToDate() {
+    return millisBehindLatest.get() == 0L;
   }
 
-  public ShardCheckpoint getCheckpoint() {
-    return checkpoint;
+  void ackRecord(KinesisRecord record) {
```

Comment:
Hm, now I'm confused, does this actually perform any acking to Kinesis or is this an in-memory operation?

Reply:
There's no acking concept in Kinesis. This ackRecord operation just creates a new in-memory  ShardCheckpoint object.